### PR TITLE
JDK-8310388: Shenandoah: Auxiliary bitmap is not madvised for THP

### PR DESCRIPTION
**Pull request by tstuefe**
See details in JBS isse.

Note that there is no actual functional difference. AUX bitmap did not use THPs before this patch and does not now either. The only difference is that before, the JVM *thought* the AUX bitmap uses THPs when in fact it did not. That caused confusion.

(All this illuminates how badly thought out the ReservedSpace API really is. We pass in page size to the constructor, but then need to commit manually; THPs actually use madvise on commit, not on reservation, so we need to pass page size in *again* to commit. Ideally, ReservedSpace should handle committing for us with the page size it saved from reservation.)

Note that this only takes care of preventing THP formation in "madvise" mode. In "always" mode, the kernel will do THP coalescation always. We could prevent it from doing so by advising against it via madvise, but that would require extension of the platform generic reservation APIs and is left for a different RFE. Ideally, nobody should use THP "always" mode.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310388](https://bugs.openjdk.org/browse/JDK-8310388): Shenandoah: Auxiliary bitmap is not madvised for THP (**Bug** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - no project role)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14953/head:pull/14953` \
`$ git checkout pull/14953`

Update a local copy of the PR: \
`$ git checkout pull/14953` \
`$ git pull https://git.openjdk.org/jdk.git pull/14953/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14953`

View PR using the GUI difftool: \
`$ git pr show -t 14953`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14953.diff">https://git.openjdk.org/jdk/pull/14953.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14953#issuecomment-1644131248)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:287]: There is a `ifdef LINUX` THP-related block in the same file, so better to match its style:

```
   size_t aux_bitmap_page_size = bitmap_page_size;
#ifdef LINUX
   // In THP "advise" mode, we refrain from advising the system to use large pages
   // since we know these commits will be short lived, and there is no reason to trash
   // the THP area with this bitmap.
   if (UseTransparedHugePages) {
     aux_bitmap_page_size = os::vm_page_size();
   }
#endif 
```

**Review by shipilev — APPROVED** [review]
Looks fine.

Does it make sense to enable `runtime/os/TestTracePageSizes.java` now?

**Review by shipilev — APPROVED** [review]

**Comment by tstuefe** [comment]
Friendly ping... still need a second review.

**Review by kdnilsen — APPROVED** [review]
Looks good to me.

**Comment by tstuefe** [comment]
Thanks Kelvin!

/integrate

**Pull request closed without merging**
